common/browser_context.go

```diff
@@ -388,52 +380,60 @@ func (b *BrowserContext) waitForEvent(event string, predicateFn goja.Callable, t
 
 	select {
 	case <-b.ctx.Done():
-		b.logger.Debugf("BrowserContext:WaitForEvent:ctx.Done", "bctxid:%v event:%q", b.id, event)
+		return nil, errors.New("test iteration ended")
 	case <-time.After(timeout):
 		b.logger.Debugf("BrowserContext:WaitForEvent:timeout", "bctxid:%v event:%q", b.id, event)
+		return nil, fmt.Errorf("waitForEvent timed out after %v", timeout)
 	case evData := <-ch:
 		b.logger.Debugf("BrowserContext:WaitForEvent:evData", "bctxid:%v event:%q", b.id, event)
-		return evData
+		return evData, nil
 	}
-	b.logger.Debugf("BrowserContext:WaitForEvent:return nil", "bctxid:%v event:%q", b.id, event)
-
-	return nil
 }
 
+// runWaitForEventHandler can work with a nil predicateFn.
+// If predicateFn is nil, it will return the response straight away.
 func (b *BrowserContext) runWaitForEventHandler(
 	ctx context.Context, evCancelFn func(), chEvHandler chan Event, out chan any, predicateFn goja.Callable,
 ) {
-	b.logger.Debugf("BrowserContext:WaitForEvent:go():starts", "bctxid:%v", b.id)
-	defer b.logger.Debugf("BrowserContext:WaitForEvent:go():returns", "bctxid:%v", b.id)
+	b.logger.Debugf("BrowserContext:runWaitForEventHandler:go():starts", "bctxid:%v", b.id)
+	defer b.logger.Debugf("BrowserContext:runWaitForEventHandler:go():returns", "bctxid:%v", b.id)
+
+	var p *Page
+	defer func() {
+		out <- p
+		close(out)
+
+		// We wait for one matching event only,
+		// then remove event handler by cancelling context and stopping goroutine.
+		evCancelFn()
+	}()
+
 	for {
 		select {
 		case <-ctx.Done():
-			b.logger.Debugf("BrowserContext:WaitForEvent:go():ctx:done", "bctxid:%v", b.id)
+			b.logger.Debugf("BrowserContext:runWaitForEventHandler:go():ctx:done", "bctxid:%v", b.id)
 			return
 		case ev := <-chEvHandler:
 			if ev.typ == EventBrowserContextClose {
```

Comment:
Reminder: This isn't being emitted, so best to remove from codebase.

Reply:
PR https://github.com/grafana/xk6-browser/pull/1045

Reply:
With the current implementation, my understanding is that if this `EventBrowserContextClose` is fired, we would be returning a `null/undefined` page to the JS context, as we do not do any verification in the `case evData := <-ch:` case in `BrowserContext.waitForEvent` select. But if this event is never fired and we will remove it, then this function will only end either in a TO or with a page being returned and passed to the given predicate function (if any). Is that right?

Reply:
> if this `EventBrowserContextClose` is fired, we would be returning a `null/undefined` page to the JS context, as we do not do any verification in the `case evData := <-ch:` case in `BrowserContext.waitForEvent` select

Yes, that's the same understanding I have of this.

> But if this event is never fired and we will remove it, then this function will only end either in a TO or with a page being returned and passed to the given predicate function (if any). Is that right?

Yes correct, or if the iteration ends.

Since `EventBrowserContextClose` isn't being emitted, feels like a good opportunity to clean things up. My main concern is that `EventBrowserContextClose`'s value is `close` and other events which are emitted for other reasons also use that same value `close` (e.g. [1](https://github.com/grafana/xk6-browser/blob/main/common/event_emitter.go#L23C39-L23C39) and [2](https://github.com/grafana/xk6-browser/blob/main/common/event_emitter.go#L32)). So I wanted to avoid any unexpected unblocks of `waitForEvent` because a `close` was emitted for an undocumented situation. I've not tested this so I could be wrong and it's safe to leave it in or complete the implementation of emitting `EventBrowserContextClose` when the `browserContext` is closed.

WDYT?